packages/ampd-sdk/src/event/event_handler.rs

```diff
@@ -0,0 +1,520 @@
+    broadcast_error_cb: fn(&[Any], H::Err) -> HandlerTaskAction,
+}
+
+#[allow(dead_code)]
+impl<H> HandlerTask<H>
+where
+    H: EventHandler,
+    <H::Event as TryFrom<Event>>::Error: Context,
+    H::Event: Display,
+{
+    pub async fn run(
+        self,
+        client: &mut impl client::Client,
+        token: CancellationToken,
+    ) -> error_stack::Result<(), Error> {
+        let task_label = self.task_label.clone();
+        let stream = self
+            .subscribe_to_stream(client, &token)
+            .await?
+            .then(|element| self.process_stream(element, &token));
+
+        pin_mut!(stream);
+        while let Some(action) = stream.next().await {
+            match action {
+                HandlerTaskAction::Continue => continue,
+                HandlerTaskAction::ContinueAndSleep(timeout) => tokio::time::sleep(timeout).await,
+                HandlerTaskAction::Abort => return Err(report!(Error::TaskAborted(task_label))),
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn subscribe_to_stream<'a>(
+        &self,
+        client: &'a mut impl client::Client,
+        token: &'a CancellationToken,
+    ) -> error_stack::Result<impl Stream<Item = error_stack::Result<Event, Error>> + 'a, Error>
+    {
+        let subscription_params = self.handler.subscription_params();
+
+        let stream = client
+            .subscribe(
+                subscription_params.event_filters,
+                subscription_params.include_block_begin_end,
+            )
+            .await
+            .change_context(Error::EventStream)?
+            .take_while(|_| !token.is_cancelled())
+            .timeout_repeating(interval(self.config.stream_timeout))
+            .map(|event| match event {
+                Ok(Ok(event)) => Ok(event),
+                Ok(Err(err)) => Err(err.change_context(Error::EventStream)),
+                Err(elapsed) => Err(Error::StreamTimeout(elapsed).into_report()),
+            });
+
+        Ok(stream)
+    }
+
+    async fn process_stream(
+        &self,
+        element: error_stack::Result<Event, Error>,
+        token: &CancellationToken,
+    ) -> HandlerTaskAction {
+        let action = self
+            .parse_event(element)
+            .and_then(|event| self.handle_event(event, token))
+            .and_then(|msgs| self.broadcast_msgs(msgs, token))
+            .await
+            .map(|_| HandlerTaskAction::Continue);
+
+        action.unwrap_or_else(|err_action| err_action)
+    }
+
+    async fn parse_event(
+        &self,
+        element: error_stack::Result<Event, Error>,
+    ) -> Result<H::Event, HandlerTaskAction> {
+        element
+            .inspect(Self::log_block_boundary)
+            .and_then(|event| {
+                H::Event::try_from(event.clone())
+                    .change_context(Error::EventConversion)
+                    .attach_printable(format!("failed to create handler event from: {event}"))
```

Comment:
```suggestion
                    .attach_printable(json!({ "event": event }))
```
Add `Serialize` to the Event derive.